### PR TITLE
8320687: sun.jvmstat.monitor.MonitoredHost.getMonitoredHost() throws unexpected exceptions when invoked concurrently

### PR DESCRIPTION
**Pull request by jaikiran**
Can I please get a review of this change which proposes to fix the issue noted in https://bugs.openjdk.org/browse/JDK-8320687?

As noted in the issue, the `sun.jvmstat.monitor.MonitoredHost.getMonitoredHost()` uses a shared instance of `java.util.ServiceLoader` to load `MonitoredHostService` services. The `ServiceLoader` class javadoc explicitly notes that it isn't thread safe. The issue at hand is caused to due using an instance of `ServiceLoader` concurrently by multiple threads.

The fix proposes to guard the usage of the shared `ServiceLoader` instance through the `monitoredHosts` object monitor. We already use that monitor when dealing with the internal cache which is populated after loading the relevant `MonitoredHostService`(s).

A new jtreg test has been introduced which always reproduces the issue without the source changes and passes with this fix.

tier1, tier2, tier3 and svc_tools tests have been run with this change and all passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8320687](https://bugs.openjdk.org/browse/JDK-8320687): sun.jvmstat.monitor.MonitoredHost.getMonitoredHost() throws unexpected exceptions when invoked concurrently (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer) ⚠️ Review applies to [669e5e5e](https://git.openjdk.org/jdk/pull/16805/files/669e5e5e0e747aac72f63d87987433d6e79d137e)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16805/head:pull/16805` \
`$ git checkout pull/16805`

Update a local copy of the PR: \
`$ git checkout pull/16805` \
`$ git pull https://git.openjdk.org/jdk.git pull/16805/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16805`

View PR using the GUI difftool: \
`$ git pr show -t 16805`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16805.diff">https://git.openjdk.org/jdk/pull/16805.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16805#issuecomment-1825191983)

### CODE REVIEW

**Comment by jaikiran** [comment]
> The fix proposes to guard the usage of the shared ServiceLoader instance through the monitoredHosts object monitor.

An alternative was to just create a new instance of `ServiceLoader` within the method implementation whenever a `MonitoredHost` was not found in the `monitoredHosts`. That works fine too. However, that will have a (relative) disadvantage that we keep loading the MonitoredHostService instances afresh and thus loses the benefits that using a single `ServiceLoader` instance provides (`ServiceLoader` instance internally maintains a cache of already loaded `Service` instances as noted in its javadoc). So I decided to stick to using a single instance of `ServiceLoader` and have its access guarded by synchronization.

Loading and instantiating `MonitoredHostService` instances within this synchronized block, I think, shouldn't cause issues because the `ServiceLoader`'s load method is passed the classloader of `sun.jvmstat.monitor.MonitoredHostService` class, which happens to be the `jdk.internal.jvmstat` JDK module. Thus it wouldn't involve unexpected user/application class instances being loaded while holding this lock.

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/GetMonitoredHost.java
Comment (AlanBateman) [inline test/jdk/sun/jvmstat/monitor/MonitoredVm/GetMonitoredHost.java:43]: This is more of a regression test for the concurrent case rather than a unit test, so I think rename to Concurrent GetMonitoredHost to make it clearer what the test is far.

**Review by AlanBateman — COMMENTED** [review]
Good sleuthing.

File: test/jdk/sun/jvmstat/monitor/MonitoredVm/GetMonitoredHost.java
Comment (jaikiran) [reply test/jdk/sun/jvmstat/monitor/MonitoredVm/GetMonitoredHost.java]: Done. I've now updated the PR to rename this new test to ConcurrentGetMonitoredHost. Test continues to pass.

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java
Comment (jaikiran) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:142]: Hello Alan,

I've updated the PR to fix the code comment section to use the existing `/* */` style.

> Can you put assert Threads.holdLock(monitoredHosts) at the top of the method, that will check that the monitor is held?

Right now, the sole usage of the `monitoredHostServiceLoader` instance is within a `synchronized (monitoredHosts) {...}` block within a method. So it wouldn't require this `assert`.

**Comment by kevinjwalls** [comment]
Yes this looks useful 8-)
sync on monitoredHosts protects monitoredHosts and monitoredHostServiceLoader.

Do we really need both synchronized(monitoredHosts) blocks -- is the first one needed, now at lines 159 -164 ?
If we don't, then you might not feel the need to create the getCachedMonitoredHost() method which separates the locking from the access you want to protect.

**Comment by jaikiran** [comment]
Hello Kevin,

> Do we really need both synchronized(monitoredHosts) blocks -- is the first one needed, now at lines 159 -164 ?
If we don't, then you might not feel the need to create the getCachedMonitoredHost() method which separates the locking from the access you want to protect.

That's a good point. I have taken your input and updated the PR to simplify the synchronized block. The test continues to pass locally (I'll re-run tier testing and svc_tools later today).

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java
Comment (AlanBateman) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:142]: > Right now, the sole usage of the `monitoredHostServiceLoader` instance is within a `synchronized (monitoredHosts) {...}` block within a method. So it wouldn't require this `assert`.

Okay, I guess part of me wonders why this field is needed in the first place. Why can't getMonitoredHost just create a ServiceLoader instead instead of trying to share between threads?

**Review by kevinjwalls — APPROVED** [review]
Great, well I like that, obviously.  Good luck with testing!

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java
Comment (jaikiran) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:142]: > > Right now, the sole usage of the `monitoredHostServiceLoader` instance is within a `synchronized (monitoredHosts) {...}` block within a method. So it wouldn't require this `assert`.
> 
> Okay, I guess part of me wonders why this field is needed in the first place. Why can't getMonitoredHost just create a ServiceLoader instead instead of trying to share between threads?


It can and that works too. That was one of the alternatives I had initially tried. I explain in this comment https://github.com/openjdk/jdk/pull/16805#issuecomment-1825194163 the reason why I thought sharing the ServiceLoader might be better. Do you think I should just use the local ServiceLoader approach instead?

Comment (AlanBateman) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:142]: > It can and that works too. That was one of the alternatives I had initially tried. I explain in this previous comment [#16805 (comment)](https://github.com/openjdk/jdk/pull/16805#issuecomment-1825194163) the reason why I thought sharing the ServiceLoader might be better. Do you think I should just use the local ServiceLoader approach instead?

I assume almost all usages just fetch the iterator just once, in which case the provider cache doesn't help. So yes, I think I would be tempted to just remove the field and make this a lot simpler.

Comment (jaikiran) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java]: Done. I've updated the PR to just use a local ServiceLoader instead of a shared one.

**Review by AlanBateman — APPROVED** [review]

File: src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java
Comment (jaikiran) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:170]: > .map(mhs -> mhs.getMonitoredHost(hostId))

That was a reasoanable suggestion and I gave it a try, but `mhs.getMonitoredHost(hostId)` throws a checked exception `MonitorException` so using this `Stream` based approach then would need additional exception catching, wrapping and throwing, then catching this wrapped exception, checking the cause and rethrowing. That then defeats the purpose of using the `Stream` approach.

I see that you have approved the PR, so I think you are OK with the current non-stream approach. I'll now trigger the tier tests and if all comes back fine, I'll go ahead and integrate this in a day or two.

Kevin, does this latest form of the PR look OK to you too?

Comment (kevinjwalls) [reply src/jdk.internal.jvmstat/share/classes/sun/jvmstat/monitor/MonitoredHost.java:170]: All good!

**Comment by jaikiran** [comment]
tier1, tier2, tier3 and svc_tools all passed with this latest state of the PR. Since serviceability area requires 2 Reviewers, could I get one more Reviewer to review this please?

**Comment by jaikiran** [comment]
>  Since serviceability area requires 2 Reviewers, could I get one more Reviewer to review this please?

I'm looking for one more Reviewer, please.

**Comment by plummercj** [comment]
> I'm looking for one more Reviewer, please.

You have 2 reviewers already.

**Comment by jaikiran** [comment]
Hello Chris, I misunderstood the 2 Reviewer rule for serviceablitiy area then. I thought it was 2 Reviewers (capital R).

I'll go ahead and integrate this shortly. Thank you all for the help and reviews.

**Pull request closed without merging**
